### ISSUE TITLE
Allow use of SIMD vector instructions (via liborc) to be toggled on/off

### ISSUE THREAD

**Issue by bkw** [lovell/sharp#172]
Hi,

here is another segfault: https://gist.github.com/bkw/b7639a397e259e297aef/

Versions:
sharp build from current master (0.9.3)
libvips build from master, jcupitt/libvips@2b9471b19d497d51999cdb29d0f3ccb99a2ebd2a
liborc-0.4.23 from http://gstreamer.freedesktop.org/src/orc/orc-0.4.23.tar.xz

I switched to another identical box and built libvips without orc there, and 1 million images later not a single segfault had happened.

//cc @jcupitt because of https://github.com/lovell/sharp/issues/151#issuecomment-70819746

**Comment by jcupitt** [comment]
@bkw, do you have any feel for how much difference orc makes to performance in your use case? If it's not too significant, perhaps we should default to orc off.

**Comment by bkw** [comment]
yes, I'll probably stick with orc disabled for now.

Because of #118 I have to do the resizing outside of sharp right now, so I guess I won't see any difference anyway.

Of course, I cannot speak for others, the convolution speedup during sharpening might still be desirable in general.

**Comment by lovell** [comment]
From @bkw's backtrace and looking at the liborc source, I think there might be a race condition when deallocating memory, but need to dig deeper. Perhaps we should report a bug upstream in liborc?

**Comment by lovell** [comment]
liborc v0.4.24 now available, which includes fixes for a couple of problems that would have resulted in a segfault - see http://cgit.freedesktop.org/gstreamer/orc/log/

**Comment by lovell** [comment]
I've decided to give liborc v0.4.24 the ultimate test. The forthcoming v0.12.0 of sharp will bundle it - looking forward to all those backtraces :)

**Comment by puzrin** [comment]
@lovel, we executed only 7000 avatars conversion with v0.12.0, and it crashed multiple times. It certainly worth to make built-in libvips without liborc. v0.11.0, compiled with `preinstall.sh` without liborc has no such problem.

Idea: you could create 2 builds, and load orc-ed one with `require('sharp').fast` as temporary solution.

I'm very interested in 0.12.+, because it has very convenient installer.

**Comment by jcupitt** [comment]
How frustrating! @puzrin, can you confirm that the crashes vanish if you disable orc? You can set the environment variable `VIPS_NOVECTOR` to stop it using orc at runtime, which might be enough. 

If it really is orc, we should probably default to orc off. I'll change vips configure.

**Comment by jcupitt** [comment]
I did some quick tests to see how much difference orc made. 

If you have a large output image, it seems to help quite a bit, since lots of pixels will be running through the output stages of the resizer. For a 10k x 10k RGB jpg resizing to 3k x 3k:

```
$ time vipsthumbnail wtc.jpg -s 3000 -o x.jpg
real	0m2.924s
user	0m10.056s
sys	0m0.072s
$ time vipsthumbnail wtc.jpg -s 3000 -o x.jpg --vips-novector
real	0m4.389s
user	0m16.216s
sys	0m0.032s
``` 

The more usual case is much smaller output images. Here, very few pixels are going through the output, so orc makes little difference:

```
$ time vipsthumbnail wtc.jpg -s 300 -o x.jpg
real	0m0.319s
user	0m0.432s
sys	0m0.012s
$ time vipsthumbnail wtc.jpg -s 300 -o x.jpg --vips-novector
real	0m0.346s
user	0m0.500s
sys	0m0.016s
```

Background: orc is the run-time vector code generation library that vips uses to make vector versions of the convolution operator. Convolution goes about 3x faster with orc enabled, convolution is used in a couple of places in the output stages of the image resizer, so overall it can make a useful difference. 

Orc is part of gstreamer, one of the main media processing libraries on linux. It's used to generate code for quite a few of the filters. It seems OK under light load, but vips, and especially sharp, are using it intensively (programs can become long and use a lot of arguments) and in a heavily threaded way. 

Back when we picked orc in 2010 there wasn't a lot of choice. These days something like OpenCL would be a better idea. It might be time to experiment with this again.

**Comment by puzrin** [comment]
> How frustrating! @puzrin, can you confirm that the crashes vanish if you disable orc? You can set the environment variable `VIPS_NOVECTOR` to stop it using orc at runtime, which might be enough.

@jcupitt I confirm. With `VIPS_NOVECTOR` - no crashes.

**Comment by lovell** [comment]
Hi @puzrin, sorry to hear you've run into this ongoing problem with liborc.

If you're able to generate one, a backtrace of the segfault should help narrow down what's causing this and provide more information to report upstream to the gstreamer maintainers.

@jcupitt I see there's a `vips_vector_set_enabled` method - perhaps sharp should expose this.

**Comment by puzrin** [comment]
@lovell i'm no strong in this area and don't know how to generate backtraces. My programmers don't know too. Need instructions how to do that. Currently we have segfaults only.

But first, i ecourage to publish sharp with stable libvips build. We have a lot of code "floating in air", and need to finish migration from im/gm to sharp. Actually, there are 2 issues:

- need migrate projects to sharp (waiting for nice installer)
- orc

**Comment by jcupitt** [comment]
I reread the orc layer and found a couple of things.

* orc now defines a set of limits for max number of instructions, constants, parameters and so on. Previously vips was relying on some rules of thumb plus a mail from the orc developer, and the rules may have changed, so it now checks against those instead.
* More orc functions now return error codes, so I check those and warn if they fail. Though I don't think this is a problem for us.
* I've put a mutex around calls to the compiler. The rest of orc looks very simple and I think can't have threading problems, but the compiler is rather hairy, it might fail if you use it from many threads.

Who knows! It might help.

https://github.com/jcupitt/libvips/commit/a0ead35bbb8510ba00003ebc0c08eb004de6dfb9

**Comment by jcupitt** [comment]
Oh, this might mean we now fail to compile against very old orc. I'll check when these extra symbols were added.

**Comment by puzrin** [comment]
```js
process.env.VIPS_NOVECTOR = 1;
```

That disables orc from node runtime. Sometime that's more convenient, than setting env variable from the shell.

Finally we can start migrating to sharp :)

**Comment by lovell** [comment]
@puzrin Great news. You'll need to ensure `process.env.VIPS_NOVECTOR = 1;` appears before the first use of `require('sharp');`.

I'm going to add a `sharp.simd(boolean)` method that will expose/call the underlying `vips_vector_set_enabled` method, defaulting to `false` for now, so the use of `VIPS_NOVECTOR` will be temporary.

@jcupitt Given the slightly worrying "various buffer overflows" warning of https://github.com/GStreamer/orc/commit/988914384acdf30eab22226d79de3541e7e6f02d your change should prevent at least one more source of segfaults, thank you.

**Comment by jcupitt** [comment]
And there are quite a few overflows they didn't stop :( ah well.

**Comment by lovell** [comment]
Commit 32c4b9e adds a new `simd` method to toggle the use of liborc, if available.
http://sharp.dimens.io/en/latest/api/#sharpsimdenable

**Comment by lovell** [comment]
v0.12.1 now available via npm - thanks everyone for all your help with this!

**Comment by puzrin** [comment]
thanks!

**Comment by gustavnikolaj** [comment]
@lovell Not sure if this is the right place, but I got this stack trace running against 0.12.0 - upgrading to 0.12.1 fixed it, but that is probably just because the disabled behaviour in 0.12.1. Running with env var VIPS_NOVECTOR=1 also fixed the problem.

```
PID 31862 received SIGSEGV for address: 0x188
./node_modules/segfault-handler/build/Release/segfault-handler.node(+0x1b70)[0x7f6e51ec7b70]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x10d10)[0x7f6e544a8d10]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/liborc-0.4.so.0(orc_code_chunk_merge+0x14)[0x7f6e412d4214]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/liborc-0.4.so.0(orc_code_chunk_free+0x55)[0x7f6e412d4285]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/liborc-0.4.so.0(orc_code_free+0x3e)[0x7f6e412e207e]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/liborc-0.4.so.0(orc_program_free+0x7b)[0x7f6e412d48eb]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(vips_vector_free+0x15)[0x7f6e433b0305]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0xcd2b6)[0x7f6e433242b6]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0xcd3a2)[0x7f6e433243a2]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0xf76c1)[0x7f6e4334e6c1]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_closure_invoke+0x190)[0x7f6e50051f10]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(+0x22e80)[0x7f6e50063e80]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_signal_emit_valist+0xdcf)[0x7f6e5006bf8f]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_signal_emit+0x82)[0x7f6e5006c252]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0x141347)[0x7f6e43398347]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_unref+0x12c)[0x7f6e50056b5c]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(vips__object_set_member+0x112)[0x7f6e433986f2]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(vips_object_set_property+0x381)[0x7f6e43398c71]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_set_valist+0x233)[0x7f6e500599b3]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_set+0xcf)[0x7f6e5005a1af]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0x13ff17)[0x7f6e43396f17]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(vips_argument_map+0x7a)[0x7f6e433981aa]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0x141327)[0x7f6e43398327]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_unref+0x12c)[0x7f6e50056b5c]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_set_valist+0x295)[0x7f6e50059a15]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_set+0xcf)[0x7f6e5005a1af]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(+0x111d7)[0x7f6e500521d7]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_signal_emit_valist+0x738)[0x7f6e5006b8f8]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_signal_emit+0x82)[0x7f6e5006c252]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(+0x141347)[0x7f6e43398347]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libgobject-2.0.so.0(g_object_unref+0x12c)[0x7f6e50056b5c]
./node_modules/express-processimage/node_modules/sharp/build/Release/../../lib/libvips.so.42(vips__object_set_member+0x112)[0x7f6e433986f2]
```

I don't know if this stack means anything to you, and if you need any further information to dig into it, please tell me, and I'll do my best to provide it.

**Comment by lovell** [comment]
@gustavnikolaj sharp v0.12.1 bundles libvips v8.1.1 so won't have the more recent https://github.com/jcupitt/libvips/commit/a0ead35bbb8510ba00003ebc0c08eb004de6dfb9 fix (and therefore disables SIMD by default). Are you able to try sharp v0.12.2+ and libvips v8.2.0+, which does, and then re-enable SIMD to see if the problem remains?

**Comment by gustavnikolaj** [comment]
@lovell Sorry about that! I did actually try it on a newer version, and didn't find your commit reverting the off-by-default behaviour. It actually works. I ran 250 iterations, with no errors, while I normally get one within the first 50 at the very most. So it looks like it is indeed fixed :-)

**Comment by jcupitt** [comment]
I think I just fell off my chair. @gustavnikolaj could you confirm that you had no errors with orc enabled? Perhaps I'm misunderstanding.

**Comment by lovell** [comment]
> "didn't find your commit reverting the off-by-default behaviour"

@gustavnikolaj The default is still off in all versions since v0.12.1 - you'll need to manually re-enable this via `sharp.simd(true)` - sorry for the confusion. Back on that chair @jcupitt.

**Comment by gustavnikolaj** [comment]
Oh! Sorry, I guess I misread it.

I'm running sharp@0.13.1 in express-processimage@4.5.3 on node 4.2. I added `sharp.simd(true)` in a local checkout of express-processimage and linked it in to the project where I had the test that could cause the segfault. I still cannot make it fail (at least not within the intervals from before).

So that's not really helpful I'm afraid. It's not a problem for me to keep running with it turned off, for now, so I don't need help - I just wanted to see if it could be of any use for you in debugging the problem. :-)

**Comment by lovell** [comment]
> "I still cannot make it fail (at least not within the intervals from before)."

That's great, John might have to hover somewhere between chair and floor.

Backtraces from sharp v0.12.2 / libvips v8.2.0 onwards with `sharp.simd(true)` will help narrow any remaining problems down further.

As an aside, I spotted a new version of liborc was released a couple of days ago so will bundle that with sharp v0.14.0 (but still disabled by default).
